velox/functions/lib/aggregates/noisy_aggregation/sketch/SfmSketch.h

```diff
@@ -0,0 +1,253 @@
+      double p = mergeRandomizedResponseProbabilities(p1, p2);
+      double normalizer = (1 - 2 * p) / ((1 - 2 * p1) * (1 - 2 * p2));
+
+      for (uint32_t i = 0; i < numberOfBits(); i++) {
+        double bit1 = bits::isBitSet(target, i) ? 1.0 : 0.0;
+        double bit2 = bits::isBitSet(source, i) ? 1.0 : 0.0;
+        double x = 1 - 2 * p - normalizer * (1 - p1 - bit1) * (1 - p2 - bit2);
+        double probability = p + normalizer * x;
+        probability = std::min(1.0, std::max(0.0, probability));
+        bits::setBit(target, i, randomizationStrategy.nextBoolean(probability));
+      }
+    }
+
+    randomizedResponseProbability_ = mergeRandomizedResponseProbabilities(
+        randomizedResponseProbability_, other.randomizedResponseProbability_);
+  }
+
+  // Make a non-private sketch private.
+  void enablePrivacy(double epsilon);
+
+  // Get a default randomization strategy.
+  static SecureRandomizationStrategy getDefaultRandomizationStrategy() {
+    return SecureRandomizationStrategy();
+  }
+
+  // Make a non-private sketch private with a randomization strategy.
+  template <typename RandomizationStrategy>
+  void enablePrivacy(
+      double epsilon,
+      RandomizationStrategy randomizationStrategy) {
+    VELOX_CHECK_NOT_NULL(
+        &randomizationStrategy, "randomizationStrategy can't be null.");
+    VELOX_CHECK(!privacyEnabled(), "privacy is already enabled.");
+    VELOX_CHECK(
+        epsilon > 0,
+        "Epsilon must be greater than zero or equal to kNonPrivateEpsilon");
+    randomizedResponseProbability_ =
+        calculateRandomizedResponseProbability(epsilon);
+
+    // Toggle each bit with a securely generated probability.
+    for (uint32_t i = 0; i < numberOfBits(); ++i) {
+      if (randomizationStrategy.nextBoolean(randomizedResponseProbability_)) {
+        bits::negateBit(bits_.data(), static_cast<uint64_t>(i));
+      }
+    }
+  }
+
+  // Get the number of bits that are set to 1.
+  int32_t countBits() const;
+
+  // Estimate the cardinality of the sketch via maximum psuedolikelihood
+  // (Newton's method).
+  uint64_t cardinality() const;
+
+  // Size of the sketch in bytes
+  size_t serializedSize() const;
+
+  // Serialize the sketch into a char array.
+  void serialize(char* out) const;
+
+  // Deserialize the sketch from a char array.
+  static SfmSketch deserialize(const char* in, HashStringAllocator* allocator);
+
+  // Probability of a 1-bit remaining a 1-bit under randomized response.
+  double onProbability() const {
+    return 1 - randomizedResponseProbability_;
+  }
+
+  uint32_t numberOfBits() const {
+    return static_cast<uint32_t>(bits_.size() * 8);
+  }
+
+  double randomizedResponseProbability() const {
+    return randomizedResponseProbability_;
+  }
+
+  uint32_t indexBitLength() const {
+    return indexBitLength_;
+  }
+
+  uint32_t precision() const {
+    return precision_;
+  }
+
+  std::vector<int8_t, facebook::velox::StlAllocator<int8_t>> bits() const {
```

Comment:
Do we need this getter? 

If we do, let's make sure to return a const & to avoid copying.